### DISCUSSION TITLE
How to handle a cancellation from client ?

### DISCUSSION THREAD

**Discussion by MrLuje** [microsoft/vs-streamjsonrpc#986] [Q&A] [answered]
Hello, I may be missing something but I don't quite understand the doc about [CancellationToken](https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/sendrequest.md#cancellation)

> A client can cancel a request by canceling a CancellationToken supplied when originally invoking the request. Be sure to call the JsonRpc.Invoke* method and overload that explicitly accepts CancellationToken (e.g. InvokeWithCancellationAsync or InvokeWithParameterObjectAsync) rather than supplying the token in the params array that must be serializable.

Let's say I use a client with `InvokeWithCancellationAsync` with `targetName`, `args` & `CancellationToken`.
```csharp
var cancellationToken = new CancellationToken();
var result = await rpc.InvokeWithCancellationAsync<int>("add", [1, 2], cancellationToken);
```
And a server:
```csharp
public class Server {
    readonly JsonRpc rpc;

    public Server(Stream sender, Stream reader)
    {
        rpc = JsonRpc.Attach(sender, reader, this);
    }

    [JsonRpcMethod("add")]
    public Task<int> Add(int a, int b) { return Task.FromResult(a + b); }
}
```

> When canceled, this CancellationToken may abort the transmission of the original request if it has not yet been transmitted. Or if the request has already been transmitted, a notification is sent to the server advising the server that the request has been canceled from the client, giving the server the opportunity to abort the invocation and return a canceled result.

I don't understand the 2nd part about the notification being sent so the server can cancel a the running task. What should I subscribe to to be notified ?

Thanks 🙏

**Comment by AArnott** [comment, marked as answer]
The server method must take `CancellationToken` as its last parameter in order to be notified of the client's request to cancel the call.

**Reply by MrLuje** [reply]
So, same client call as my example but server becomes the following :
```csharp
public class Server {
    readonly JsonRpc rpc;

    public Server(Stream sender, Stream reader)
    {
        rpc = JsonRpc.Attach(sender, reader, this);
    }

    [JsonRpcMethod("add")]
    public Task<int> Add(int a, int b, CancellationToken ct) { 
        // properly use ct here...

        return Task.FromResult(a + b);
    }
}
```

**Reply by AArnott** [reply]
Yes.
And BTW, you can avoid using explicit `JsonRpcMethod` attributes for every method just to change capitalization by following [this doc](https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/recvrequest.md#special-method-names).

**Reply by MrLuje** [reply]
Good to know, thanks
